src/jit/lowerxarch.cpp

```diff
@@ -1174,13 +1174,13 @@ void Lowering::TreeNodeInfoInitShiftRotate(GenTree* tree)
     else
     {
         MakeSrcContained(tree, shiftBy);
-    }
 
-    // Codegen of shift oper sets ZF and SF flags.
-    // Note that Rotate Left/Right instructions don't set ZF and SF flags.
-    if (tree->OperIsShift())
-    {
-        tree->gtFlags |= GTF_ZSF_SET;
+        // Codegen of shift oper sets ZF and SF flags provided shift count is not zero.
+        // Note that Rotate Left/Right instructions don't set ZF and SF flags.
+        if (tree->OperIsShift() && (shiftBy->gtIntConCommon.IconValue() != 0))
```

Comment:
If we haven't already narrowed the constant value to [0..63]

Then you probably need to to handle shifts of 64,128,192, etc.. 
As those values may also behave the same as a a shift of zero on the hardware.